Respect policy list expiration

Snawoot opened the pull request:
[Policy list specification](https://github.com/EFForg/starttls-everywhere/blob/master/RULES.md) requires generator to take action when list is expired:
> If the file has ceased being regularly updated for any reason, and the policy file has expired, the MTA should fall-back to opportunistic TLS for e-mail delivery, and the system operator should be alerted.

This PR implements such behavior. Actions taken on expired list:

1. Generate empty config with single comment about reason.
2. Outputs warning message to stderr. Basically, it should make CRON alarm go off if user runs list update and config generation unattended.

Signed-off-by: Vladislav Yarmak <vladislav@vm-0.com>

Snawoot commented:
Hi! Any updates?

sydneyli commented on starttls-policy/starttls_policy/configure.py:
```diff
@@ -49,6 +51,11 @@ def manual_instructions(self):
+    def expired_warning(self):
+        """Warns user about policy list expiration."""
+        six.print_("\nACTION REQUIRED: policy list expired. Probably, policy update ceased. "
+                   "Generating empty config.\n", file=sys.stderr)
```
Let's change this text to something like:
"\nACTION REQUIRED: your policy list at {config_location} has expired! Generating empty config.\n
Check to see whether your update mechanism (cronjob, systemd timer) is working.\n"

sydneyli commented on starttls-policy/starttls_policy/configure.py:
```diff
@@ -79,6 +86,9 @@ class PostfixGenerator(ConfigGenerator):
     def _generate(self, policy_list):
+        if util.is_expired(policy_list.expires):
+            self.expired_warning()
+            return "# Policy list is outdated. Falling back to opportunistic encryption."
```
Since the expired warning should probably be the same across different config generators, let's move this check & report up to `ConfigGenerator.generate`?

However, the resulting "fallback" may change depending on the MTA, so maybe we can add an abstract `_generate_expired_fallback` that returns the "empty fallback" contents for that particular MTA.

I'm open to other ideas on this too. What do you think?

Snawoot commented:
Pull request updated, all requested changes applied.

sydneyli commented on starttls-policy/starttls_policy/configure.py:
```diff
@@ -49,10 +55,25 @@ def manual_instructions(self):
 
+    @abc.abstractmethod
+    def _generate_expired_fallback(self, policy_list):
+        """Creates configuration file fox expired policy list.
```
"file for expired policy list" ?

Snawoot replied:
Fixed

sydneyli commented on starttls-policy/starttls_policy/configure.py:
```diff
@@ -49,10 +55,25 @@ def manual_instructions(self):
                     mta_name=self.mta_name,
                     instructions=self._instruct_string()))
 
+    def expired_warning(self):
```
Prefix function with an underscore to indicate it's meant for internal use only

Snawoot replied:
I noticed `manual_instructions` method is named as public so I decided to do same with `expired_warning` in order to provide same "visibility" for somewhat similar methods. Is it wrong?

sydneyli replied:
Great question!
From a high level, we wanted any consumer of a ConfigGenerator to be able to do two things:
 - Generate a configuration file for an MTA
 - Install that configuration file to that MTA

Which correspond to the `generate` and `manual_instructions`, respectively. The `manual_instructions` name is a bit of a misnomer; it was originally `install`, but was changed since we were concerned about modifying people's MTA configs.

So although `manual_instrcutions` and `expired_warning` both do similar things under the hood (writing to a std fd), their intended use is different-- a consume of ConfigGenerator should ideally never have to bring up an expiry warning since we expect `generate` to do that.

Snawoot replied:
Ok, got it! Fixed.

Snawoot commented:
I noticed moving expiration logic has broken tests because they test generator via private method and skip expiration logic. So I modified tests:
* Created necessary infrastructure to conduct tests of generator via public methods with writing config file.
* Removed code duplication introduced by earlier versions of this pull request and established testcase parameterization across tested configs.

Please also check tests part.

sydneyli commented:
By the way, I may be splitting the code portion out of the repository today-- just a heads up-- and migrating some of the PRs over.

Snawoot commented on starttls-policy/starttls_policy/tests/configure_test.py:
```diff
@@ -44,7 +48,41 @@ def test_manual_instructions(self):
+    def __exit__(self, exc_type, exc_value, traceback):
+        os.remove(self._policy_filename)
+        os.rmdir(self._tmpdir)
+
```
`mock` is a great tool, but I doubt about it's purpose here. Mocking actual file IO allows us to run tests clearly, without interacting with real FS. But creation of temporary directory with file is not something we can't afford.

In general, best practice is to mock as less as possible, resort to mock only when counterpart our code tested against cannot be efficiently represented by real instance.

Mock solution has following benefit over real FS write: it avoids usage of real FS. But it has drawbacks:
* Reader will **always** read `read_data` with same type of string `mock_open` was initialized. Consider [following examples](https://gist.github.com/Snawoot/33b797bad8265a78f1071626d41de559). So, `mock_open` is imperfect.
* Reading of file by a program may produce different results depending on method of reading. For example, there are no reasons why `policy.json` can't have byte order mark in the beginning of file and to handle such case we have to use special BOM-aware codec. Mocked file defined by a string doesn't allow us to conduct tests against real files.
* Mocking requires test to have some knowledge of internal structure of program in order it can be properly patched, and this knowledge has to be maintained. What if we will use some faster json module which will load json file directly?

BTW, when our project will become popular (I believe!) and policy list will be big enough to not fit in memory, we probably will have to reconsider parsing of json file anyway in order to reduce memory footprint required to generate policy. Maybe generator-like json parser like `ijson` will do.

I think if we will need to mock file IO, we can return to it in future. But for now it is unnecessary.

Snawoot commented:
Ok, moving discussion to new repo
